sql: make global indexes a new doc and add more contents (#22017)
pingcap/docs: global-indexes.md

---
title: Global Indexes
summary: Learn the use cases, advantages, usage, working principles, and limitations of TiDB global indexes.
---

# Global Indexes

> Before the introduction of global indexes, TiDB created a local index for each partition, leading to [a limitation](/partitioned-table.md#partitioning-keys-primary-keys-and-unique-keys) that primary keys and unique keys had to include the partition key to ensure data uniqueness. Additionally, when querying data across multiple partitions, TiDB needed to scan the data of each partition to return results.

Reviewer: ```suggestion
Before introducing global indexes, TiDB created a local index for each partition, meaning one local index per partition. This indexing approach had [a limitation](/partitioned-table.md#partitioning-keys-primary-keys-and-unique-keys) that primary keys and unique keys had to include all the partition keys to ensure global uniqueness of data. In addition, when a query needed to access data across multiple partitions, TiDB had to scan the data of each partition to return results.
```

To address these issues, TiDB introduces the global indexes feature in [v8.3.0](/releases/release-8.3.0.md). A global index covers the data of the entire table with a single index, allowing primary keys and unique keys to maintain global uniqueness without including all partition keys. Moreover, global indexes can access index data across multiple partitions in a single operation instead of looking up the local index for each partition, significantly improving query performance for non-partitioned keys. <!--Starting from v9.0.0, non-unique indexes can also be created as global indexes.-->

## Advantages

Global indexes significantly improve query performance, enhance indexing flexibility, and reduce the cost of data migration and  modifying applications.

### Improved query performance

> Global indexes greatly enhance the efficiency of queries involving non-partition columns. When a query involves a non-partition column, a global index can quickly locate the relevant data, avoiding full table scans across all partitions. This dramatically reduces the number of Coprocessor (cop) tasks, which is especially beneficial in scenarios with a large number of partitions.

Author: ```suggestion
Global indexes greatly enhance the efficiency of queries involving non-partitioning columns. When a query involves a non-partitioning column, a global index can quickly locate the relevant data, avoiding full table scans across all partitions. This dramatically reduces the number of Coprocessor (cop) tasks, which is especially beneficial in scenarios with a large number of partitions.
```

> In benchmark tests using sysbench `select_random_points`, performance improves by up to 53 times when the table contains 100 partitions.

Reviewer: ```suggestion
Benchmark tests show that when a table contains 100 partitions, performance in the sysbench `select_random_points` scenario improves by up to 53 times.
```

### Enhanced indexing flexibility

Global indexes remove the restriction that unique keys in partitioned tables must include all partitioning columns. This provides greater flexibility in index design. You can now create indexes based on actual query patterns and business logic, rather than being constrained by the partitioning scheme. This flexibility not only improves query performance but also supports a wider range of application requirements.

### Reduced cost for data migration and modifying applications

> Global indexes significantly simplify adjustments for data migration and modifying application. Without global indexes, you might need to modify partitioning schemes or rewrite queries to work around indexing limitations. With global indexes, such changes are unnecessary, reducing both development and maintenance overhead.

Reviewer: ```suggestion
During data migration and application modification, global indexes can significantly reduce the amount of additional adjustment work required. Without global indexes, you might need to change the partitioning scheme or rewrite SQL queries to work around index limitations. With global indexes, these modifications can be avoided, reducing both development and maintenance costs.
```

> For example, when migrating a table from an Oracle database to TiDB, because Oracle supports global indexes, some tables might contain unique indexes that do not include partitioning columns. Before TiDB introduced global indexes, you had to modify the table schema to comply with TiDB's partitioning rules. Now, TiDB supports global indexes, you can simply define those indexes as global during migration, keeping schema behavior consistent with Oracle and greatly reducing migration costs.

Reviewer: ```suggestion
For example, when migrating a table from an Oracle database to TiDB, you might encounter unique indexes that do not include partitioning columns, because Oracle supports global indexes. Before TiDB introduced global indexes, you had to modify the table schema to comply with TiDB's partitioning rules. Now, TiDB supports global indexes. When you migrate data, you can simply define those indexes as global, keeping schema behavior consistent with Oracle and greatly reducing migration costs.
```

> #### Limitations of global indexes

Author: ```suggestion
## Limitations of global indexes
```

- If a global index is added to an expression column, or a global index is also a prefix index (for example `UNIQUE KEY idx_id_prefix (id(10)) GLOBAL`), you need to collect statistics manually for this global index.

## Feature evolution

- **Before v7.6.0**: TiDB only supports local indexes on partitioned tables. This means that unique keys on partitioned tables have to include all columns in the partition expression. Queries that do not use the partition key have to scan all partitions, resulting in degraded query performance.
- **[v7.6.0](https://docs.pingcap.com/tidb/stable/release-7.6.0)**: Introduces the [`tidb_enable_global_index`](/system-variables.md#tidb_enable_global_index-new-in-v760) system variable to enable global indexes. However, at that time the feature is still under development and is not recommended for production use.
- **[v8.3.0](https://docs.pingcap.com/tidb/stable/release-8.3.0)**: Global indexes are released as an experimental feature. You can explicitly create a global index using the `GLOBAL` keyword when defining an index.

> - **[v8.4.0](https://docs.pingcap.com/tidb/stable/release-8.4.0)**: The global indexes feature becomes generally available (GA). You can create global indexes directly using the `GLOBAL` keyword without setting the `tidb_enable_global_index` system variable. From this version onward, the system variable is deprecated and fixed to `ON`, meaning global indexes are enabled by default.

Reviewer: ```suggestion
- **[v8.4.0](https://docs.pingcap.com/tidb/stable/release-8.4.0)**: The global indexes feature becomes generally available (GA). You can create global indexes directly using the `GLOBAL` keyword without setting the `tidb_enable_global_index` system variable. Starting from this version, the system variable is deprecated and its value is fixed to `ON`, meaning global indexes are enabled by default.
```

## Global indexes vs. local indexes

The following diagram shows the differences between global indexes and local indexes.

<img src="https://github.com/hfxsd/docs/blob/global-index-best-practices/media/global-index-vs-local-index.png" alt="Global Index vs. Local Index" width="60%" height="60%"/>

**Scenarios for global indexes**:

- **Infrequent data archiving**: For example, in the healthcare industry, some business data must be retained for up to 30 years. Data is often partitioned monthly, creating 360 partitions at once, with very few `DROP` or `TRUNCATE` operations. In such scenarios, global indexes are more suitable, providing cross-partition consistency and improved query performance.
- **Queries that require cross-partition data**: When queries need to access data across multiple partitions, global indexes can avoid full scans across all partitions and enhance query efficiency.

**Scenarios for local indexes**:

- **Frequent data archiving**: If data archiving operations are frequent and queries are mostly confined to a single partition, local indexes can offer better performance.

> - **Partition exchange requirements**: In industries like banking, processed data might first be written to a regular table and, after verification, exchanged into a partitioned table to minimize performance impact. In this case, local indexes are preferred, because enabling global indexes disables the partition exchange functionality on the table.

Reviewer: ```suggestion
- **Frequent data archiving**: If data archiving operations occur frequently and most queries are limited to a single partition, local indexes can offer better performance.
- **Use of partition exchange**: In industries like banking, processed data might be written to a regular table first and then exchanged into a partitioned table after verification, to minimize performance impact on the partitioned table. In this case, local indexes are preferred because once a global index is used, the partitioned table no longer supports partition exchange.
```

## Global indexes vs. clustered indexes

Due to the underlying principles of clustered indexes and global indexes, a single index cannot serve as both a clustered index and a global index. However, these two types of indexes provide different performance optimizations for different query scenarios. When you need to leverage the benefits of both, you can add the partition columns to the clustered index while also creating a global index that does not include the partition columns.

Suppose you have the following table structure:

```sql
CREATE TABLE `t` (
  `id` int DEFAULT NULL,
  `ts` timestamp NULL DEFAULT NULL,
  `data` varchar(100) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY RANGE (UNIX_TIMESTAMP(`ts`))
(PARTITION `p0` VALUES LESS THAN (1735660800)
 PARTITION `p1` VALUES LESS THAN (1738339200)
 ...)
```

> In the preceding `t` table, the values in the `id` column are unique. To optimize both point queries and range queries, you can define a clustered index in the table creation statement as `PRIMARY KEY(id, ts)` and a global index without the partition column as `UNIQUE KEY id(id)`. This way, point queries based on `id` will use the global index `id` and choose a `PointGet` execution plan, while range queries will use the clustered index. The clustered index requires one less table lookup compared to the global index, improving query efficiency.

Author: ```suggestion
In the preceding `t` table, the values in the `id` column are unique. To optimize both point queries and range queries, you can define a clustered index in the table creation statement as `PRIMARY KEY(id, ts)` and a global index without the partitioning column as `UNIQUE KEY id(id)`. This way, point queries based on `id` will use the global index `id` and choose a `PointGet` execution plan, while range queries will use the clustered index. The clustered index requires one less table lookup compared to the global index, improving query efficiency.
```

> The modified table structure is as follows:

Reviewer: ```suggestion
The modified table schema is as follows:
```

## Usage

To create a global index, add the `GLOBAL` keyword in the index definition.

> **Note:**
>
> Global indexes affect partition management. Executing `DROP`, `TRUNCATE`, or `REORGANIZE PARTITION` operations will trigger updates to the table-level global indexes. This means that these DDL operations only return after the corresponding global index updates are completed, which might increase execution time.

```sql
CREATE TABLE t1 (
    col1 INT NOT NULL,
    col2 DATE NOT NULL,
    col3 INT NOT NULL,
    col4 INT NOT NULL,
    UNIQUE KEY uidx12(col1, col2) GLOBAL,
    UNIQUE KEY uidx3(col3),
    KEY idx1(col1) GLOBAL
)
PARTITION BY HASH(col3)
PARTITIONS 4;
```

In the preceding example, the unique index `uidx12` and the non-unique index `idx1` become global indexes, while `uidx3` remains a regular unique index.

Note that a clustered index cannot be a global index. For example:

```sql
CREATE TABLE t2 (
    col1 INT NOT NULL,
    col2 DATE NOT NULL,
    PRIMARY KEY (col2) CLUSTERED GLOBAL
) PARTITION BY HASH(col1) PARTITIONS 5;
```

```
ERROR 1503 (HY000): A CLUSTERED INDEX must include all columns in the table's partitioning function
```

> A clustered index cannot simultaneously serve as a global index. This is because if a clustered index were global, the table would no longer be partitioned. The keys of a clustered index are at the partition level, while global indexes operate at the table level, creating a conflict. If you need to set the primary key as a global index, you must explicitly define it as a non-clustered index, for example: 

Reviewer: ```suggestion
A clustered index cannot also serve as a global index. This is because if a clustered index is global, the table would no longer be partitioned. The key of a clustered index is the key for the partition-level row data, while a global index is defined at the table level, creating a conflict. If you need to make the primary key a global index, you must explicitly define it as a non-clustered index. For example:
```

| test         | t1         |          1 | idx1     |            1 | col1        |     NULL |               | NULL       |        3 | YES        | NO        |         1 |
+--------------+------------+------------+----------+--------------+-------------+----------+---------------+------------+----------+------------+-----------+-----------+
3 rows in set (0.00 sec)
```

When partitioning a regular table or repartitioning a partitioned table, you can update indexes to be either global indexes or local indexes as needed.

For example, the following SQL statement repartitions table `t1` based on column `col1` and updates the global indexes `uidx12` and `idx1` to local indexes, while updating the local index `uidx3` to a global index. `uidx3` is a unique index on column `col3`. To ensure the uniqueness of `col3` across all partitions, `uidx3` must be a global index. `uidx12` and `idx1` are indexes on column `col1` and can be either global or local indexes.

```sql
ALTER TABLE t1 PARTITION BY HASH (col1) PARTITIONS 3 UPDATE INDEXES (uidx12 LOCAL, uidx3 GLOBAL, idx1 LOCAL);
```

## Working mechanism

> This section explains the working mechanism of global indexes, including their design concept and implementation.

Reviewer: ```suggestion
This section explains the working mechanism of global indexes, including their design principles and implementation.
```

### Design concept

In TiDB partitioned tables, the key prefix of a local index is the partition ID, while the key prefix of a global index is the table ID. This design ensures that the data of a global index is stored continuously on TiKV, reducing the number of RPC requests when querying the index.

```sql
CREATE TABLE `sbtest` (
  `id` int(11) NOT NULL,
  `k` int(11) NOT NULL DEFAULT '0',
  `c` char(120) NOT NULL DEFAULT '',
  KEY idx(k),
  KEY global_idx(k) GLOBAL
) partition by hash(id) partitions 5;
```

> Take the preceding table structure as an example: `idx` is a local index, and `global_idx` is a global index. The data of `idx` is distributed across 5 different ranges, such as `PartitionID1_i_xxx` and `PartitionID2_i_xxx`. Whereas the data of `global_idx` is concentrated in a single range (T`ableID_i_xxx`).

Author: ```suggestion
Take the preceding table structure as an example: `idx` is a local index, and `global_idx` is a global index. The data of `idx` is distributed across 5 different ranges, such as `PartitionID1_i_xxx` and `PartitionID2_i_xxx`. Whereas the data of `global_idx` is concentrated in a single range (`TableID_i_xxx`).
```

The table structure used in the tests is as follows:

```sql
CREATE TABLE `sbtest` (
  `id` int(11) NOT NULL,
  `k` int(11) NOT NULL DEFAULT '0',
  `c` char(120) NOT NULL DEFAULT '',
  `pad` char(60) NOT NULL DEFAULT '',
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,
  KEY `k_1` (`k`)
  /* Key `k_1` (`k`, `c`) GLOBAL */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
/* Partition by hash(`id`) partitions 100 */
/* Partition by range(`id`) xxxx */
```

The workload SQL is as follows:

```sql
SELECT id, k, c, pad
FROM sbtest1
WHERE k IN (xx, xx, xx)
```

Range Partition (100 partitions):

| Table type                                                            | Concurrency 1 | Concurrency 32 | Concurrency 64 | Average RU |
| --------------------------------------------------------------------- | ------------- | -------------- | -------------- | ---------- |
| Clustered non-partitioned table                                       | 225           | 19,999         | 30,293         | 7.92       |
| Clustered table range partitioned by PK                               | 68            | 480            | 511            | 114.87     |

> | Clustered table range partitioned by PK, with Global Index on `k`,`c` | 207           | 17,798         | 27,707         | 11.73      |

Author: ```suggestion
| Clustered table range partitioned by PK, with Global Index on `k`, `c` | 207           | 17,798         | 27,707         | 11.73      |
```

---
title: Global Indexes
summary: Introduce the use cases, advantages, usage, working principles, and limitations of TiDB global indexes.
---

# Global Indexes

Before the introduction of global indexes, TiDB created a local index for each partition, leading to [a limitation](/partitioned-table.md#partitioning-keys-primary-keys-and-unique-keys) that primary keys and unique keys had to include the partition key to ensure data uniqueness. Additionally, when querying data across multiple partitions, TiDB needed to scan the data of each partition to return results.

To address these issues, TiDB introduces the global indexes feature in [v8.3.0](/releases/release-8.3.0.md). A global index covers the data of the entire table with a single index, allowing primary keys and unique keys to maintain global uniqueness without including all partition keys. Moreover, global indexes can access index data across multiple partitions in a single operation instead of looking up the local index for each partition, significantly improving query performance for non-partitioned keys. <!--Starting from v9.0.0, non-unique indexes can also be created as global indexes.-->

> 

Author: ```suggestion
```
